src/esp/gfx/Shader.cpp

```diff
@@ -0,0 +1,119 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "Shader.h"
+
+#include <Magnum/Shaders/Flat.h>
+#include <Magnum/Shaders/Phong.h>
+#include <Magnum/Shaders/Shaders.h>
+
+#include "esp/gfx/PrimitiveIDShader.h"
+#ifdef ESP_BUILD_PTEX_SUPPORT
+#include "esp/gfx/PTexMeshShader.h"
+#endif
+#include "esp/gfx/Drawable.h"
+#include "esp/gfx/RenderCamera.h"
+
+namespace esp {
+namespace gfx {
+
+namespace {
+// TODO: don't hardcode this
+std::unique_ptr<Magnum::GL::AbstractShaderProgram> shaderProgramFactory(
+    const ShaderConfiguration& cfg) {
+  std::unique_ptr<Magnum::GL::AbstractShaderProgram> shaderProgram;
+  switch (cfg.type) {
+    case ShaderType::INSTANCE_MESH_SHADER: {
+      return std::make_unique<gfx::PrimitiveIDShader>();
+    } break;
+#ifdef ESP_BUILD_PTEX_SUPPORT
+    case ShaderType::PTEX_MESH_SHADER: {
+      return std::make_unique<gfx::PTexMeshShader>();
+    } break;
+#endif
+
+    case ShaderType::COLORED_SHADER: {
+      return std::make_unique<Magnum::Shaders::Flat3D>(
+          Magnum::Shaders::Flat3D::Flag::ObjectId);
+    } break;
+
+    case ShaderType::VERTEX_COLORED_SHADER: {
+      return std::make_unique<Magnum::Shaders::Flat3D>(
+          Magnum::Shaders::Flat3D::Flag::ObjectId |
+          Magnum::Shaders::Flat3D::Flag::VertexColor);
+    } break;
+
+    case ShaderType::TEXTURED_SHADER: {
+      return std::make_unique<Magnum::Shaders::Flat3D>(
+          Magnum::Shaders::Flat3D::Flag::ObjectId |
+          Magnum::Shaders::Flat3D::Flag::Textured);
+    } break;
+
+    case ShaderType::COLORED_SHADER_PHONG: {
+      shaderProgram = std::make_unique<Magnum::Shaders::Phong>(
+          Magnum::Shaders::Phong::Flag::ObjectId, 3 /*lights*/);
+    } break;
+
+    case ShaderType::VERTEX_COLORED_SHADER_PHONG: {
+      shaderProgram = std::make_unique<Magnum::Shaders::Phong>(
+          Magnum::Shaders::Phong::Flag::ObjectId |
+              Magnum::Shaders::Phong::Flag::VertexColor,
+          3 /*lights*/);
+    } break;
+
+    case ShaderType::TEXTURED_SHADER_PHONG: {
+      shaderProgram = std::make_unique<Magnum::Shaders::Phong>(
+          Magnum::Shaders::Phong::Flag::ObjectId |
+              Magnum::Shaders::Phong::Flag::DiffuseTexture,
+          3 /*lights*/);
+    } break;
+
+    default:
+      return nullptr;
+      break;
+  }
+
+  /* Default setup for Phong, shared by all models */
```

Comment:
Will all Shaders of same shaderProgram type be collapsed into one setup? (e.g. COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG different only in config, not separately sub-classed)

Reply:
Yep, currently we should just have Flat, PrimitiveID, Phong, and PtexMesh ShaderProgram classes, and COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG will only differ in config

Reply:
Yep, currently we should just have Flat, PrimitiveID, Phong, and PtexMesh ShaderProgram classes, and COLORED_SHADER_PHONG and TEXTURED_SHADER_PHONG will only differ in config